File: app/src/main/java/com/zaclimon/acetv/data/RoomContentPersistence.java
package com.zaclimon.acetv.data;

import android.content.Context;

import com.zaclimon.xipl.model.AvContent;
import com.zaclimon.xipl.persistence.ContentPersistence;

import java.util.ArrayList;
import java.util.List;

/**
 * Concrete implementation of {@link ContentPersistence} applicable for the Room persistence library.
 *
 * @author zaclimon
 * Creation date: 25/03/18
 */

public class RoomContentPersistence implements ContentPersistence {

    /**
     * {@link AvContentDao} used to access the data from {@link AceTvDatabase}
     */
    private AvContentDao mContentDao;

    public RoomContentPersistence(Context context) {
        mContentDao = AceTvDatabase.getInstance(context).avContentDao();
    }

    @Override
    public void insert(AvContent avContent) {
        mContentDao.bulkInsert(new RoomAvContent(avContent));
    }

    @Override
    public void insert(List<AvContent> avContents) {
        RoomAvContent[] tempArray = new RoomAvContent[avContents.size()];
        mContentDao.bulkInsert(convertToRoom(avContents).toArray(tempArray));
    }

    @Override
    public void deleteCategory(String category) {
        mContentDao.deleteCategory(category);
    }

    @Override
    public void deleteAll() {
        mContentDao.deleteAll();
    }

    @Override
    public List<AvContent> getAll() {
        List<RoomAvContent> tempContents = mContentDao.getAll();
        return (convertFromRoom(tempContents));
    }

    @Override
    public List<AvContent> getFromCategory(String category, boolean isAlphabeticallySorted) {
        List<RoomAvContent> tempContents = mContentDao.getFromCategory(category);
        return (convertFromRoom(tempContents));
    }

    @Override
    public List<AvContent> getFromTitle(String title, boolean isAlphabeticallySorted) {
        List<RoomAvContent> tempContents = mContentDao.getFromTitle(title);
        return (convertFromRoom(tempContents));
    }

    @Override
    public long size() {
        return (mContentDao.size());
    }

    @Override
    public long size(String category) {
        return (mContentDao.size(category));
    }

    private List<RoomAvContent> convertToRoom(List<AvContent> contents) {
        List<RoomAvContent> tempList = new ArrayList<>();

        for (AvContent content : contents) {
            tempList.add(new RoomAvContent(content));
        }
        return (tempList);
    }

    private List<AvContent> convertFromRoom(List<RoomAvContent> contents) {

        List<AvContent> tempList = new ArrayList<>();

        for (RoomAvContent content : contents) {
            String title = content.getTitle();
            String logo = content.getLogo();
            String group = content.getGroup();
            String contentLink = content.getContentLink();
            String contentCategory = content.getContentCategory();
            int id = content.getId();
            tempList.add(new AvContent(title, logo, group, contentCategory, contentLink, id));
        }
        return (tempList);
    }

}

File: app/src/main/java/com/zaclimon/tsutaeru/data/RoomContentPersistence.kt
package com.zaclimon.tsutaeru.data

import android.content.Context
import com.zaclimon.xipl.model.AvContent
import com.zaclimon.xipl.persistence.ContentPersistence

/**
 * Concrete implementation of [ContentPersistence] applicable for the Room persistence library.
 *
 * @property contentDao the DAO that will be used for interaction between the application and the Room database
 *
 * @author zaclimon
 */
class RoomContentPersistence(context: Context) : ContentPersistence {

    private val contentDao: AvContentDao = TsutaeruDatabase.getInstance(context).avContentDao()

    override fun insert(avContent: AvContent) {
        contentDao.bulkInsert(RoomAvContent(avContent))
    }

    override fun insert(avContents: MutableList<AvContent>) {
        contentDao.bulkInsert(*convertToRoom(avContents).toTypedArray())
    }

    override fun deleteCategory(category: String) {
        contentDao.deleteCategory(category)
    }

    override fun deleteAll() {
        contentDao.deleteAll()
    }

    override fun getAll(): MutableList<AvContent> {
        return convertFromRoom(contentDao.getAll()).toMutableList()
    }

    override fun getFromTitle(title: String, isAlphabeticallySorted: Boolean): MutableList<AvContent> {
        return convertFromRoom(contentDao.getFromTitle(title)).toMutableList()
    }

    override fun getFromCategory(category: String, isAlphabeticallySorted: Boolean): MutableList<AvContent> {
        return convertFromRoom(contentDao.getFromCategory(category)).toMutableList()
    }

    override fun size(): Long {
        return contentDao.size()
    }

    override fun size(category: String): Long {
        return contentDao.size(category)
    }

    /**
     * Converts a given list of [AvContent]s to be usable for [RoomAvContent]
     *
     * @param contents the list containing [AvContent]s
     * @return the list of converted [RoomAvContent]s
     */
    private fun convertToRoom(contents: List<AvContent>): List<RoomAvContent> {
        val tempList = mutableListOf<RoomAvContent>()

        for (content in contents) {
            tempList.add(RoomAvContent(content))
        }
        return tempList
    }

    /**
     * Converts a given list of [RoomAvContent]s to be usable for [AvContent]
     *
     * @param contents the list containing [RoomAvContent]
     * @return the list of converted [AvContent]s
     */
    private fun convertFromRoom(contents: List<RoomAvContent>): List<AvContent> {
        val tempList = mutableListOf<AvContent>()

        for (content in contents) {
            val id = content.contentId
            val title = content.contentTitle
            val logo = content.contentLogo
            val group = content.contentGroup
            val link = content.contentLink
            val category = content.contentCategory
            tempList.add(AvContent(title, logo, group, link, category, id))
        }
        return tempList
    }
}